backend/poll.go

```diff
@@ -0,0 +1,195 @@
+	Questions []string
+}
+
+//holds stored Poll, PollAnswers is an array where index is the question and the value is the vote count
+type Poll struct {
+	PollAnswers []int
+}
+
+//holds a boolean value used for storing the ClientId
+type ClientId struct {
+	voted bool
+}
+
+var questions []string
+var pollQuestions PollQuestions
+
+func InitPollSample() {
+	questions = []string{"Penguins", "Ostriches", "Kiwis", "Wekas"}
+	pollQuestions = PollQuestions{questions}
+	http.HandleFunc(POLL_SAMPLE_PATH+"submit", submitPoll)
+	RegisterSample(CATEGORY_SAMPLE_TEMPLATES+"/poll", handlePoll)
+}
+
+func handlePoll(w http.ResponseWriter, r *http.Request, page Page) {
+	page.Render(w, pollQuestions)
+}
+
+func createPollResult(answers []int, message string) PollResult {
+	results := make([]PollEntryResult, len(questions))
+	totalAnswers := 0
+	 for _, num := range answers {
+			 totalAnswers += num
+	 }
+	hundredDividedByTotalVotes := float64(100)/float64(totalAnswers)
+	for questionIndex, votes := range answers {
+		results[questionIndex] = PollEntryResult{votes,
+			make([]int, int(hundredDividedByTotalVotes * float64(votes))),
+			questions[questionIndex]}
+	}
+	return PollResult{results, message}
+}
+
+func submitPoll(w http.ResponseWriter, r *http.Request) {
+	EnableCors(w, r)
+	SetContentTypeJson(w)
+	context := appengine.NewContext(r)
+	pollForm, error := parsePollForm(w, r, context)
+	if error != nil {
+		handleError(error, w)
+		return
+	}
+	pollResult, error := calculatePollResults(w, r, context, pollForm)
+	if error != nil {
+		handleError(error, w)
+		return
+	}
+	json.NewEncoder(w).Encode(pollResult)
+}
+
+func parsePollForm(w http.ResponseWriter, r *http.Request, context context.Context) (PollForm, error) {
+	answer, answerErr := parseNotEmptyFormValue(r, "answer")
+	answerId, _ := strconv.Atoi(answer)
+	clientId, clientIdErr := parseNotEmptyFormValue(r, "clientId")
+
+	pollForm := PollForm{clientId, answerId}
+
+	error := parseFormErrors([]error{answerErr, clientIdErr})
+	return pollForm, error
+}
+
+func calculatePollResults(w http.ResponseWriter, r *http.Request, ctx context.Context, pollForm PollForm) (PollResult, error) {
+	//get the poll
+	pollKey := datastore.NewKey(ctx, "Poll", "poll", 0, nil)
+	var poll Poll
+	var pollExistingAnswers = make([]int, len(questions))
+	err := datastore.Get(ctx, pollKey, &poll)
+	//if there is an existing poll, get all the answers
+	if err == nil && (len(poll.PollAnswers) == len(questions)) {
+		pollExistingAnswers = poll.PollAnswers
+	}
+	//check if the user has already voted by checking the user id cookie existance,
+	cookie, err := r.Cookie(POLL_COOKIE_NAME)
+	if err == nil {
+		//return the existing poll answers and message to let the user know that has already voted
+		return createPollResult(pollExistingAnswers, ALREADY_VOTED_MESSAGE), nil
+	}
+		//cookies don't work on Safari when accessing the cdn, check if the user has already
+		//voted by checking the client id
+		clientId := pollForm.ClientId
```

Comment:
Oh, I see. The name `POLL_USER_ID` makes it sounds as if the cookie is specific to the poll, rather than an existing user id of potentially long standing. (Also, it actually [gets set to the value of clientId](https://github.com/ampproject/amp-by-example/blob/master/backend/poll.go#L165)--that could be just a random number, right?)